proposals/p5337.md

```diff
@@ -0,0 +1,1021 @@
+    prevent the overlap.
+-   It would reduce the ability to compose libraries.
+-   We want to support examples like this interaction between `Array`, `As`, and
+    `ImplicitAs`.
+
+Note that prioritizing the final `impl` over the non-final `impl` (even when the
+non-final `impl` has a preferred type structure) produces the desired result in
+the example above. By writing `final` we ensure that `Convert` in `ImplicitAs`
+always matches `Convert` in `As`, when they are both defined.
+
+The next question is what to do if there are two overlapping final `impl`s. The
+restrictions on
+[which libraries can define a final `impl`](/docs/design/generics/details.md#libraries-that-can-contain-a-final-impl)
+mean there are at most two files that can contain two final `impl` definitions
+that both match a particular (type, interface) query: the one defining the root
+type and the one defining the root interface. The only way to have the final
+`impl`s in different files is if there is a
+[blanket impl](/docs/design/generics/details.md#blanket-impl-declarations) of
+the interface in the file defining the interface and the other in the file
+defining the type. This latter file will have to transitively depend on the file
+defining the interface in order to write the `impl` definition, and so will be
+able to see any `final` blanket impls.
+
+This means there may be files that will only see the final `impl` definitions in
+the same file as the interface definition. Those final `impl` definitions will
+be assumed to apply whenever they match, and so they must be prioritized over
+any other overlapping final `impl` defined in the file with the type. We
+therefore adopt the rule that final `impl` definitions in the library with the
+type are only selected if they match and no final `impl` definitions in the
+library with the interface do (in the case those are different files).
+
+> _**Concern**_: This creates a situation where a final `impl` defined in the
+> file with the type may not be selected for some queries that it matches, which
+> may be surprising.
+
+#### Overlapping `final` impls
+
+If there is any other overlap between final `impl` definitions, it must be
+within the same file. We have two tools here we could use to select between the
+candidates: type structure and
+[`impl_priority`/`match_first` blocks](/docs/design/generics/details.md#prioritization-rule).
+(We will use the placeholder term of "`match_first` blocks" to refer to these
+without asserting how they will be spelled.)
+
+The decision here is to only use `match_first` blocks, not type structure to
+choose between overlapping final `impl` definitions in the same file, for a few
+reasons:
+
+-   It would be surprising to prefer something with a specified type over a
+    blanket `impl` when they are in the same file, as type structure would
+    dictate, since we are saying the blanket `impl` will have priority when they
+    are in different files.
```

Comment:
But we said the blanket one has priority only because it's final, and I think the current design prevents the specialized one from being final (which I think is a problem but mentioned above...). In this case to have priority the specialization with the more specified type would have to be final so I am not sure this is surprising here.